File: backend/app/services/bookshelf_service.py
from sqlalchemy.orm import Session
from typing import List, Optional
from ..models.bookshelf import Bookshelf, BookshelfBook

class BookshelfService:
    def __init__(self, db: Session):
        self.db = db

    def create_bookshelf(self, user_id: int, name: str, description: Optional[str] = None, is_public: bool = False) -> Bookshelf:
        db_bookshelf = Bookshelf(
            user_id=user_id,
            name=name,
            description=description,
            is_public=1 if is_public else 0
        )
        self.db.add(db_bookshelf)
        self.db.commit()
        self.db.refresh(db_bookshelf)
        return db_bookshelf

    def get_user_bookshelves(self, user_id: int) -> List[Bookshelf]:
        return self.db.query(Bookshelf).filter(Bookshelf.user_id == user_id).all()

    def add_book_to_bookshelf(self, bookshelf_id: int, book_id: int) -> BookshelfBook:
        db_bookshelf_book = BookshelfBook(
            bookshelf_id=bookshelf_id,
            book_id=book_id
        )
        self.db.add(db_bookshelf_book)
        self.db.commit()
        self.db.refresh(db_bookshelf_book)
        return db_bookshelf_book 

File: frontend/src/services/bookshelf-service.ts
const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';

export interface Bookshelf {
    id: number;
    name: string;
    description?: string;
    is_public: boolean;
    user_id: number;
    created_at: string;
    updated_at: string;
    book_count?: number;
}

export interface CreateBookshelfRequest {
    name: string;
    description?: string;
    is_public: boolean;
}

class BookshelfService {
    private baseUrl: string;

    constructor(baseUrl: string) {
        this.baseUrl = baseUrl;
    }

    private async request<T>(endpoint: string, options: RequestInit = {}): Promise<T> {
        const url = `${this.baseUrl}${endpoint}`;
        const config: RequestInit = {
            headers: {
                'Content-Type': 'application/json',
                ...options.headers,
            },
            ...options,
        };

        try {
            const response = await fetch(url, config);

            if (!response.ok) {
                throw new Error(`HTTP error! status: ${response.status}`);
            }

            return await response.json();
        } catch (error) {
            console.error('Bookshelf service request failed:', error);
            throw error;
        }
    }

    async getUserBookshelves(): Promise<Bookshelf[]> {
        return this.request<Bookshelf[]>('/bookshelves');
    }

    async createBookshelf(data: CreateBookshelfRequest): Promise<Bookshelf> {
        const formData = new FormData();
        formData.append('name', data.name);
        if (data.description) {
            formData.append('description', data.description);
        }
        formData.append('is_public', data.is_public.toString());

        return this.request<Bookshelf>('/bookshelves', {
            method: 'POST',
            body: formData,
        });
    }

    async addBookToBookshelf(bookshelfId: number, bookId: number): Promise<void> {
        return this.request<void>(`/bookshelves/${bookshelfId}/books/${bookId}`, {
            method: 'POST',
        });
    }

    async removeBookFromBookshelf(bookshelfId: number, bookId: number): Promise<void> {
        return this.request<void>(`/bookshelves/${bookshelfId}/books/${bookId}`, {
            method: 'DELETE',
        });
    }
}

export const bookshelfService = new BookshelfService(API_BASE_URL); 